cam_diagnostics/cam_diagnostics.F90

```diff
@@ -0,0 +1,211 @@
+module cam_diagnostics
+
+   use ccpp_kinds, only:  kind_phys
+   use ccpp_constituent_prop_mod, only: ccpp_constituent_prop_ptr_t
+
+   implicit none
+   private
+   save
+
+   public :: cam_state_diagnostics_init ! init routine
+   public :: cam_state_diagnostics_run  ! main routine
+   public :: cam_tend_diagnostics_init ! init routine
+   public :: cam_tend_diagnostics_run  ! main routine
+
+   character(len=65) :: const_std_names(4) = &
+   (/'water_vapor_mixing_ratio_wrt_moist_air_and_condensed_water       ', &
+     'cloud_liquid_water_mixing_ratio_wrt_moist_air_and_condensed_water', &
+     'rain_mixing_ratio_wrt_moist_air_and_condensed_water              ', &
+     'cloud_ice_mixing_ratio_wrt_moist_air_and_condensed_water         '/)
+
+   character(len=6) :: const_diag_names(4) = (/'Q     ', &
+                                               'CLDLIQ', &
+                                               'CLDICE', &
+                                               'RAINQM'/)
```

Comment:
we're looping over the constituents object, yes. But this is a mapping of the standard name to the diagnostic name (which is not contained within the object). Not all of these will necessarily be "used". Does that make any sense?

Reply:
This still feels "fragile" to me, as the diagnostic name index may not match the constituent index.  Right now in ZM, I am assuming that the tendency array maps one-to-one to the constituent array.  I would naively think that the diagnostic array would also use the same indexing.  Let me know if we should have a discussion on this.

Reply:
Hi @cacraigucar I'm not quite following. Is the problem that someone could modify one of the lists (`const_std_names` or `const_diag_names`) and then they'd be off? To clarify, when I'm looping over the constituents, I don't expect them to be in this exact order. This is just a mapping between standard name and diagnostic name (which is not contained in the properties object). Let me know if I'm still missing something!

Reply:
thanks @cacraigucar and @jimmielin - I've opened https://github.com/ESCOMP/atmospheric_physics/issues/137 so to keep track of the things that we can simplify/clean up once we have access to the full constituents object. That will solve both concerns (we won't need these mapping arrays anymore)

Reply:
> 
> My immediate thought was to assign these into a Nx2 array instead of two N-sized arrays but a quick Google didn't reveal if Fortran had syntax like this, which would prevent typing errors as much as possible:

@jimmielin, the standard way of initializing a multi-rank array would look something like this:

```
   character(len=24) :: tarr(2,3) = reshape((/ &
        'water_vapor_mixing_ratio', 'Q                       ',          &
        'cloud_liquid_water      ', 'CLDLIQ                  ',          &
        'cloud_ice_water         ', 'CLDICE                  ' /), (/ 2, 3 /))
```

Reply:
Thanks @gold2718! I suppose this could help avoid the two lists from being mismatched. I was hoping to avoid the `reshape` since the array dimensions would need to be typed twice (in the definition and in the reshape call) and probably very far away from each other.

Reply:
When I do this, I always use a parameter for each dimension size.
The only example of this I found in CAM code is from the very ancient (well before my time) src/physics/camrt/radae.F90

Newer code, such as src/dynamics/se/dycore/fvm_consistent_se_cslam.F90 does have a parameter for the changeable dimension which I think addresses your main safety concern.

Reply:
Thanks! I like this in `fvm_consistent_se_cslam.F90`, which I guess is what you're referring to: it's as close to safe and 'automatic' as we can get before the clean-up issue #137 is addressed:
```fortran
    integer, parameter :: num_area=5
    integer, dimension(num_area*4), parameter :: idy_shift_tmp = (/-1, 0, 0, 0,-1,&  !iside=1
                                                                   -1,-1, 0, 1, 1,&  !iside=2
                                                                    1, 0, 0, 0, 1,&  !iside=3
                                                                    1, 1, 0,-1,-1/)  !iside=4

    integer, dimension(num_area,4), parameter :: idy_shift = RESHAPE(idy_shift_tmp,(/num_area,4/))
```

Reply:
While this conversation has been good, my concern was not actually about there being two arrays, but rather about the necessity of having them at all.  I had a google chat with Courtney which resulted in #137 and will eventually eliminate the need for these arrays.

Reply:
Thanks @cacraigucar, my apologies for misunderstanding the original issue. Until #137 becomes a more permanent fix the current code looks good!

Reply:
No apology needed.  Nobody understood what my concern was until I met with Courtney and explained it in person!